pageserver/src/tenant/size.rs

```diff
@@ -44,6 +44,129 @@ struct TimelineInputs {
+// Update { lsn: 0/169AD58, command: BranchFrom(Some(1)), timeline_id: 2 },
+// Update { lsn: 0/169AD58, command: BranchFrom(Some(2)), timeline_id: 3 },
+//
+fn sort_updates_in_tree_order(updates: Vec<Update>) -> anyhow::Result<Vec<Update>> {
+    let mut sorted_updates = Vec::with_capacity(updates.len());
+
+    info!("Sorting updates in tree order {:?}", updates);
+
+    let mut known_timelineids = HashSet::new();
+    let mut i = 0;
+    while i < updates.len() {
+        let curr_upd = &updates[i];
+
+        if let Command::BranchFrom(parent_id) = curr_upd.command {
+            // root timeline, branches from None
+            let parent_id = match parent_id {
+                Some(parent_id) if known_timelineids.contains(&parent_id) => {
+                    // we have already processed ancestor
+                    // process this BranchFrom Update normally
+                    known_timelineids.insert(curr_upd.timeline_id);
+                    sorted_updates.push(*curr_upd);
+                    i += 1;
+                    continue;
+                }
+                None => {
+                    // root timeline, branches from None
+                    known_timelineids.insert(curr_upd.timeline_id);
+                    sorted_updates.push(*curr_upd);
+                    i += 1;
+                    continue;
+                }
+                Some(parent_id) => parent_id,
+            };
+
+            let mut j = i;
+
+            // we have not processed ancestor yet.
+            // there is a chance that it is at the same Lsn
+            if !known_timelineids.contains(&parent_id) {
+                info!("Found possibly orphan branch {:?}", curr_upd);
+
+                let mut curr_lsn_branchfroms: HashMap<TimelineId, Vec<(TimelineId, usize)>> =
+                    HashMap::new();
+
+                // inspect all branchpoints at the same lsn
+                while j < updates.len() && updates[j].lsn == curr_upd.lsn {
+                    let lookahead_upd = &updates[j];
+                    j += 1;
+
+                    if let Command::BranchFrom(lookahead_parent_id) = lookahead_upd.command {
+                        match lookahead_parent_id {
+                            Some(lookahead_parent_id)
+                                if !known_timelineids.contains(&lookahead_parent_id) =>
+                            {
+                                // we have not processed ancestor yet
+                                // store it for later
+                                curr_lsn_branchfroms
+                                    .entry(lookahead_parent_id)
+                                    .and_modify(|e| {
+                                        e.push((lookahead_upd.timeline_id, j));
+                                    })
+                                    .or_insert_with(|| vec![(lookahead_upd.timeline_id, j)]);
+                            }
+                            _ => {
```

Comment:
Nit, but prefer non-exhaustive matches.
```suggestion
                            None => {
```

Reply:
this _ actually covers not only None,
Some branch above is conditional

Reply:
Ah missed it! Resolving.